src/keys.c

```diff
@@ -2966,8 +2977,30 @@ int DeriveKeys(WOLFSSL* ssl)
             wc_ShaFinal(sha, shaOutput);
 
             XMEMCPY(md5Input + SECRET_LEN, shaOutput, SHA_DIGEST_SIZE);
-            wc_Md5Update(md5, md5Input, SECRET_LEN + SHA_DIGEST_SIZE);
-            wc_Md5Final(md5, keyData + i * MD5_DIGEST_SIZE);
+            ret = wc_Md5Update(md5, md5Input, SECRET_LEN + SHA_DIGEST_SIZE);
+            if (ret != 0) {
+            #ifdef WOLFSSL_SMALL_STACK
+                XFREE(shaOutput, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(md5Input,  NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(shaInput,  NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(keyData,   NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(md5,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
+                XFREE(sha,       NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            #endif
+                return ret;
+            }
+            ret = wc_Md5Final(md5, keyData + i * MD5_DIGEST_SIZE);
+            if (ret != 0) {
+            #ifdef WOLFSSL_SMALL_STACK
+                XFREE(shaOutput, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
Too much code is added here.  Notice that other failures drop down to bottom to free in one place.